airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/DestinationTaskLauncher.kt

```diff
@@ -81,13 +87,53 @@ class DefaultDestinationTaskLauncher(
     private val processRecordsTaskFactory: ProcessRecordsTaskFactory,
     private val processBatchTaskFactory: ProcessBatchTaskFactory,
     private val closeStreamTaskFactory: CloseStreamTaskFactory,
-    private val teardownTaskFactory: TeardownTaskFactory
+    private val teardownTaskFactory: TeardownTaskFactory,
+    private val exceptionHandler: DestinationTaskLauncherExceptionHandler
 ) : DestinationTaskLauncher {
     private val log = KotlinLogging.logger {}
 
     private val runTeardownOnce = AtomicBoolean(false)
     private val batchUpdateLock = Mutex()
 
+    class SyncTaskWrapper(
+        private val exceptionHandler: DestinationTaskLauncherExceptionHandler,
+        private val innerTask: Task,
+    ) : Task {
+        override suspend fun execute() {
+            try {
+                innerTask.execute()
+            } catch (t: Throwable) {
+                exceptionHandler.handleSyncFailure(t)
```

Comment:
is the expectation that `handleSync/StreamFailure` never themselves throw an exception?

(... or are they supposed to _rethrow_ the exception?)

also, nit: do we really want to catch throwable here? or just exception

Reply:
It should just be exception, not throwable.

I would expect exceptions in the handlers themselves to be un-handleable and just crash the sync.

I'm still deciding on the question of whether to have it be expected to rethrow (in the next PR). That was pending my dumb question in the internal channel.

Reply:
> exceptions in the handlers

makes sense. maybe dumb question - coroutines give us that behavior for freeish, right? b/c exceptions eventually naturally bubble up to the main thread

(i.e. unlike in plain java/kotlin, where uncaught exceptions on other threads don't do anything unless you handle  them explicitly)

Reply:
Yes, we get that for free, but only because of the way I'm doing the TaskRunner.

There's a separate way to dispatch things async like this:

```
val scope = CoroutineScope()

// ...

fun doStuff() {
  // ...
  scope.launch { someAsyncTask() }
  // ...
}
```

In the above case, `doStuff` will return just fine, but any exceptions thrown will silently cancel the scope.

You can inspect it later and determine that it failed, but AFAIK you can't get the exception out (unless the underlying task sticks it somewhere).